File: go/kata/structures/trie.go
package structures

import (
	"strings"
	"unicode"
)

type Trie struct {
	root *TrieNode
}

type TrieNode struct {
	char     rune
	isWord   bool
	children map[rune]*TrieNode
}

func NewTrie() *Trie {
	return &Trie{
		root: &TrieNode{children: buildChildrenMap()},
	}
}

func newTrieNode(char rune) *TrieNode {
	return &TrieNode{
		char:     char,
		isWord:   false,
		children: buildChildrenMap(),
	}
}

func (t *Trie) Insert(word string) {
	word = strings.ToLower(word)
	curr := t.root

	for _, char := range word {
		if curr.children[char] == nil {
			curr.children[char] = newTrieNode(char)
		}

		curr = curr.children[char]
	}

	curr.isWord = true
}

func (t *Trie) Find(prefix string) []string {
	buffer := make([]rune, 50)
	bufferIdx := 0

	// The first thing to do is to find the node that will act as the root
	curr := t.root
	for _, chr := range prefix {
		next := curr.children[unicode.ToLower(chr)]

		if next == nil {
			return []string{}
		}

		curr = next
		buffer[bufferIdx] = chr
		bufferIdx++
	}

	bufferIdx--

	// Now that we've found the root node, we make a depth-first search to find all possible words starting from it
	words := make([]string, 0, 10)
	find(curr, &buffer, bufferIdx, &words)

	return words
}

func find(curr *TrieNode, buffer *[]rune, bufferIdx int, words *[]string) int {
	if curr == nil {
		return -1
	}

	(*buffer)[bufferIdx] = curr.char
	bufferIdx++

	if curr.isWord {
		*words = append(*words, string((*buffer)[0:bufferIdx]))
	}

	for _, child := range curr.children {
		idx := find(child, buffer, bufferIdx, words)

		if idx != -1 {
			bufferIdx = idx - 1
		}
	}

	return bufferIdx
}

func buildChildrenMap() map[rune]*TrieNode {
	return make(map[rune]*TrieNode)
}

File: typescript/kata/trie.ts
type TrieNode = {
  value: string;
  isWord: boolean;
  children: Map<string, TrieNode>
}

class Trie {
  private root: TrieNode;

  constructor() {
    this.root = { children: new Map() } as TrieNode;
  }

  insert(word: string) {
    word = word.toLowerCase();

    let curr = this.root;
    for (const char of word) {
      let node = curr.children.get(char);

      if (node === undefined) {
        node = { value: char, children: new Map() } as TrieNode;
        curr.children.set(char, node);
      }

      curr = node as TrieNode;
    }

    curr.isWord = true;
  }

  find(prefix: string): string[] {
    const buffer: string[] = [];
    let bufferIdx = 0;

    let curr = this.root;
    for (const char of prefix) {
      const next = curr.children.get(char.toLowerCase());

      if (next === undefined) {
        return [];
      }

      curr = next;
      buffer[bufferIdx++] = char;
    }

    bufferIdx--;

    const words: string[] = [];
    this.findDepthFirst(curr, buffer, bufferIdx, words);

    return words;
  }

  private findDepthFirst(curr: TrieNode | undefined, buffer: string[], bufferIdx: number, words: string[]): number {
    if (curr === undefined) {
      return -1;
    }

    buffer[bufferIdx++] = curr.value;

    if (curr.isWord) {
      words.push(buffer.slice(0, bufferIdx).join(''));
    }

    for (const child of curr.children.values()) {
      const idx = this.findDepthFirst(child, buffer, bufferIdx, words);

      if (idx !== -1) {
        bufferIdx = idx - 1;
      }
    }

    return bufferIdx;
  }

}

const words = [
  "archimedes",
  "ari",
  "aristóteles",
  "ariosvaldo",
  "ariel",
  "antes",
  "avião",
  "aliás",
  "após",
  "apoio",
  "apoiar",
  "casinha",
  "casa",
  "casaco",
  "caso",
  "camarão",
  "cadastro",
  "comida",
  "coisa",
  "mari",
  "mariana",
];

const trie = new Trie();
for (const word of words) {
  trie.insert(word);
}

console.log(trie.find("ari"));
console.log(trie.find("mari"));
console.log(trie.find("Ma"));
console.log(trie.find("ca"));
